embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/FragmentBreadcrumbDataSource.kt

```diff
@@ -1,90 +1,82 @@
-         * by [RemoteConfig].
-         */
-        private const val DEFAULT_VIEW_STACK_SIZE = 20
-    }
-
-    internal val fragmentStack: MutableList<FragmentBreadcrumb> = Collections.synchronizedList(ArrayList<FragmentBreadcrumb>())
-
-    fun startFragment(name: String?): Boolean {
-        if (name == null) {
-            return false
-        }
-        synchronized(this) {
-            if (fragmentStack.size >= DEFAULT_VIEW_STACK_SIZE) {
-                return false
+    /**
+     * Called when a fragment is started.
+     */
+    fun startFragment(name: String?): Boolean = startSpan(
+        inputValidation = { !name.isNullOrEmpty() },
+        captureAction = {
+            val data = FragmentBreadcrumb(checkNotNull(name), clock.now()).toSpanEventData()
+            startSpan(data)?.apply {
+                fragmentSpans[name] = this
             }
-            return fragmentStack.add(FragmentBreadcrumb(name, clock.now(), 0))
         }
-    }
+    )
 
-    fun endFragment(name: String?): Boolean {
-        if (name == null) {
-            return false
-        }
-        var start: FragmentBreadcrumb
-        val end = FragmentBreadcrumb(name, 0, clock.now())
-        synchronized(this) {
-            val crumbs = filter(fragmentStack) { crumb: FragmentBreadcrumb -> crumb.name == name }
-            if (crumbs.isEmpty()) {
-                return false
-            }
-            start = crumbs[0]
-            fragmentStack.remove(start)
+    /**
+     * Called when a fragment is ended.
+     */
+    fun endFragment(name: String?): Boolean = stopSpan(
+        inputValidation = { !name.isNullOrEmpty() },
+        captureAction = {
+            val span = fragmentSpans.remove(name)
+            span?.endFragmentSpan(clock.now())
         }
-        end.setStartTime(start.getStartTime())
-        store.tryAddBreadcrumb(end)
-        return true
-    }
+    )
 
     /**
-     * Close all open fragments when the activity closes
+     * Called when the activity is closed (and therefore all fragments are assumed to close).
      */
     fun onViewClose() {
-        if (!configService.breadcrumbBehavior.isActivityBreadcrumbCaptureEnabled()) {
-            return
-        }
-        if (fragmentStack.size == 0) {
-            return
-        }
-        val ts = clock.now()
-        synchronized(fragmentStack) {
-            for (fragment in fragmentStack) {
-                fragment.endTime = ts
-                store.tryAddBreadcrumb(fragment)
-            }
-            fragmentStack.clear()
+        fragmentSpans.forEach { (_, span) ->
+            stopSpan(
+                inputValidation = NoInputValidation,
+                captureAction = { span.endFragmentSpan(clock.now()) }
+            )
         }
     }
 
-    override fun cleanCollections() {
-        store.cleanCollections()
-        fragmentStack.clear()
+    override fun FragmentBreadcrumb.toSpanEventData(): SpanEventData = SpanEventData(
```

Comment:
Ah, I think I got confused here - `SpanEventData` here is being used to represent the input to create a span as well as to create a SpanEvent (for the custom breadcrumb data source).

Maybe we shouldn't overload this name/class but instead have different data classes for each use case?